### ISSUE TITLE
Understanding GitHub permissions

### ISSUE THREAD

**Issue by consideRatio** [yuvipanda/github-app-user-auth#5]
There are two times that the end user via a GitHub website, and I don't yet understand what roles they have etc. To understand that is quite relevant for anyone being an admin of a hub that provides this if the admin gets questions from users.

# Step 1 - When an end user installs the github application, granting it specific permissions to various repos

This can be triggered by visiting https://github.com/apps/hub-jupytearth-org-github-integ

![image](https://user-images.githubusercontent.com/3837114/151056467-00dfec6f-d42f-4a74-832d-876caa07b69d.png)

# Step 2 - When an end user authorize the application to "act on your behalf"

This can be triggered by running `github-app-auth-user`, getting a device code, and opening https://github.com/login/device

![image](https://user-images.githubusercontent.com/3837114/151056588-00c965c1-71f7-4b08-8444-caaf2416c18b.png)
![image](https://user-images.githubusercontent.com/3837114/151056643-9d84ee64-7394-446f-9565-4445ec1e03ec.png)
![image](https://user-images.githubusercontent.com/3837114/151056192-dd8219f0-34a1-4ca4-8bcf-13ceea38e1f4.png)

**Comment by yuvipanda** [comment]
So the actual permissions granted are specified by the admin when they *create* the app itself. Step 1 grants permissions to the app to act on the repos, while step 2 grants permission for the app to act *on behalf of* the user on those particular repos - if they users don't have access to the repo they still can't push. So the effective permission requires that:

1. GitHub app is created with correct scopes (push, pull)
2. GitHub app is installed in the *repos* where we want end users to eventually push / pull
3. Users authorize the app to act on their behalf, so users still need to (independently) have permissions to these repos.

I hope this helps! Should also be more clearly documented

**Comment by consideRatio** [comment]
Thanks @yuvipanda I think you summarized it as comprehensible as something this complex can be summarized!

Hmmm, so there is like two separate distinctions:
- Does the user trust the app to act with your credentials?
- Does the github repo owners trust the app to make changes for users who trusts the app?

I don't yet understand this as clearly as I hope to, but I do understand it better thanks to your help @yuvipanda!

I think what I look to understand more clearly are the security implications in various scenarios, for example if a GitHub app developer is malicious, and what steps then are needed for that malicious GitHub app developer needs to get done to do harm - and how great that harm will be - and such questions.

**Comment by yuvipanda** [comment]
@consideRatio oooh, great question. I *think* the developer of the app could possibly write to the repos even without user authentication :| https://github.com/yuvipanda/git-credential-helpers relies on that to work with nbgitpuller on private repositories. We should make that clear.

**Comment by consideRatio** [comment]
Ah now i understand that project:
1 create an app
2 create a key (in app settings) to act as the app
3 declare permissions for app
4 install app on org/repo
5 use app credentials (via git cred. helper) to do what you want

Okay, so the app has permissions by itself.

Hmmmm, but in this case (github-app-user-auth) a user knowing the client-id (public) can use the app to generate credentials which is the intersection of the app credentials and the user credentials, and with github-app-user-auth, those stay with the end user rather than to be used by some server doing automation etc?

**Comment by yuvipanda** [comment]
@consideRatio yeah, so if you *just* know the client id, you can use it only in conjugation with user credentials to get access to repos with intersection of these permissions. You need access to the GitHub app's private key (not used in this repo) to get access without user authorization. So the presence of the github app ID acts purely as a constraint on what scopes the user's permissions are reduced to.

**Comment by yuvipanda** [comment]
No key is created here - only the (public) client ID is used.
